### PR TITLE
WIP: Apply Attributes

### PR DESCRIPTION
**Pull request by kooldev**
Changes handling of internal attributes like this:

1. Validation of internal attributes is deferred to a point where the compiler is not active (this enables use of auto-loading when constant ASTs have to be evaluated).
2. Attributes are validated by calling their constructors. This avoids the need to re-implement argument validation in a C callback. (It would be ideal to reuse PHP's parameter parsing code without actually calling the constructor here...)
3. Internal attributes may declare an "apply" callback that is called whenever attributes are compiled or loaded from opcache.

### CODE REVIEW

**Comment by kooldev** [comment]
@nikic This (WIP) PR is an attempt to refactor PHP internal attribute validation in a way that makes it easy to validate arguments (by using "standard" constructors). It enables auto-loading in AST args by deferring calls to a point in time where the compiler is not active. In addition it also provides an optional "apply" callback that is called whenever an internal attribute is compiled (or loaded from Opcache). Do you think this approach is worthwhile or should I just trash it? 🤔

Code related to Opcache is a PoC. Traversing all classes / functions whenever a script is loaded from Opcache is not a great solution. It might be possible to store all required apply calls in a flat list / `HashTable` in `persistent_script` instead, that should be a bit faster.

In general I am trying to fill in the gaps that we never closed while implementing internal attributes in PHP 8. With "New in initializers" making it into PHP 8.1 this seems to be long-overdue...

**Comment by nikic** [comment]
I'm not sure I really understand what problem these changes are trying to solve. Why do you want to instantiate internal attributes during class declaration, even though we don't do the same (and explicitly don't want to do the same) for userland attributes?

File: Zend/zend_language_scanner.l [hunk 1]
```diff
@@ -632,6 +633,10 @@ static zend_op_array *zend_compile(int type)
 	zend_arena_destroy(CG(ast_arena));
 
 	CG(in_compilation) = original_in_compilation;
+	
+	if (!CG(in_compilation) && CG(apply_attributes)) {
+		zend_apply_attributes();
```
Comment (nikic) [inline Zend/zend_language_scanner.l:638]: From a semantics perspective, I would have expected this to happen when the class is actually declared, not when the file is compiled. However, the class could be at least partially immutable at that point, so I can see why you do it here instead.

**Comment by kooldev** [comment]
Actually I don't really want to instantiate attributes. The only thing this is supposed to do is to validate attribute arguments against arginfo declared by the constructor. I would love to omit creating objects if I could validate this in another way.

The basic idea is that since we cannot fully validate attribute AST args during compilation (due to possible auto-loading) this step should be deferred. Another caveat is that we need to revalidate AST args when loading from Opcache because they may refer to values / classes in other scripts that might have changed in the meantime.

**Comment by nikic** [comment]
@kooldev As this is only about internal attributes, not userland ones, why is autoloading relevant? All internals classes should be loaded by the time we compile userland classes with attributes. (Well, I guess you could have an autoloader that dl()'s an extension, but I don't have interest in supporting such cases.)

**Comment by kooldev** [comment]
@nikic What about a situation like this one:

```php
#[Attribute(OtherClass::TARGET)]
class MyAttribute { }
```

This is an internal attribute but it requires auto-loading (if `OtherClass` is not loaded yet) to resolve it's `flags` argument. This is an artificial example but I think it illustrates how auto-loading can interfere with internal attributes.

If you run this on PHP 8.0 it yields a fatal error `Class "OtherClass" not found` because `validate_attribute` cannot resolve the value.

**Comment by nikic** [comment]
Thanks, I see what you mean now.

It's worth noting that allowing autoloading during compilation isn't really a problem technically (the compiler is reentrant nowadays), but there is the opcache problem you mention (could reference different class on next execution).

However, delaying attribute application causes another problem with opcache. Say we have an attribute like `#[Deprecated]` that wants to make a modification to the class/method/function by adding a `ZEND_ACC_DEPRECATED` flag to it. Functions in opcache are immutable, so it wouldn't be possible to do so when loading from opcache. Classes/methods are initially lazy-loaded (and modification is possible at that point), but once they are in the inheritance cache they're also immutable and we have the same problem.

As such, I'm not sure this can really work. I think it's pretty important that internal attribute apply happens at a point where it is legal to modify the class/function, and that means it has to happen before it enters opcache. But if it happens before entering opcache, it can't have per-request behavior.

**Comment by kooldev** [comment]
How about splitting this into an "early" `apply` and a per-request `bind`? Things like `Deprecated` could use `apply` to modify elements before they enter opcache. Other things (like the `Attribute` flags validation) could use `bind` instead because they don't have to modify things before hitting Opcache.

One of the things I had in mind with moving `apply` to per-request was that it would allow for something like Python's decorators to be implemented sometime in the future.

**Comment by nikic** [comment]
@kooldev Well, decorators would be one of the things that require modifying the class, so I don't think that would work with per-request either. Or is what you have in mind something like replacing the method with a shim implementation in `apply` and then determining the actual function the shim is supposed to call in `bind`? That would work I guess, though at that point one could also lazily determine it on first call of the shim.

**Comment by kooldev** [comment]
@nikic You are right, I forgot about functions being immutable once a per-request action is possible. The few benefits of supporting autoloading in args are probably not worth the added complexity...

I think will put this aside and not bother with it anymore. Thank you anyway for looking into it.

**Pull request closed without merging**